proposals/push-constants.md

```diff
@@ -0,0 +1,73 @@
+# Subgroups
+
+Status: **Draft**
+
+Last modified: 2024-04-30
+
+Issue: #75
+
+# Requirements
+
+No special requirements.
+
+# WGSL
+
+
+## Address Spaces
+
+| Address space | Sharing among invocations | Default access mode | Notes |
+| --- | --- | --- | --- |
+| `push_constant` | Invocations in [shader stage](https://www.w3.org/TR/WGSL/#shader-stages) | [read](https://www.w3.org/TR/WGSL/#access-read) | For [uniform buffer](https://www.w3.org/TR/WGSL/#uniform-buffer) variables exclude [array types](https://www.w3.org/TR/WGSL/#array-types) variable or [structure types](https://www.w3.org/TR/WGSL/#struct-types) variable contains [array types](https://www.w3.org/TR/WGSL/#array-types) attributes |
```

Comment:
>Is the uniform buffer reference a copy paste error?

No, I think `uniform buffer` variable defined in wgsl aligns with what I want with `push constant`. I think `push constant variables` should be [host-shareable](https://www.w3.org/TR/WGSL/#host-shareable) [constructible](https://www.w3.org/TR/WGSL/#constructible) type, and must satisfy the [address space layout constraints](https://www.w3.org/TR/WGSL/#address-space-layout-constraints).

>Separately, are arrays excluded to avoid Vulkan's default uniform access requirements? 

Not only Vulkan, but also DX12. DX12 says _Arrays are not permitted in constant buffers that get mapped onto root constants since dynamic indexing in the root signature space is not supported_ (https://learn.microsoft.com/en-us/windows/win32/direct3d12/using-constants-directly-in-the-root-signature)